addons/http_routing/models/ir_http.py

```diff
@@ -321,15 +337,22 @@ def _add_dispatch_parameters(cls, func):
         if request.routing_iteration == 1:
             context = dict(request.context)
             path = request.httprequest.path.split('/')
-            langs = [code for code, _ in request.env['res.lang'].get_installed()]
             is_a_bot = cls.is_a_bot()
-            cook_lang = request.httprequest.cookies.get('frontend_lang')
+
+            lang_ids = cls._get_languages()
             nearest_lang = not func and cls.get_nearest_lang(path[1])
-            preferred_lang = ((cook_lang if cook_lang in langs else False)
-                              or (not is_a_bot and cls.get_nearest_lang(request.lang))
-                              or cls._get_default_lang().code)
+            cook_lang = request.httprequest.cookies.get('frontend_lang')
+            cook_lang = lang_ids.filtered(lambda l: l.code == cook_lang)
 
-            request.lang = context['lang'] = nearest_lang or preferred_lang
+            # TODO: replace this change by lazy_property back on WebRequest but in
+            #       http_routing to be able to browse the record set
+            preferred_lang = (cook_lang
+                              or (not is_a_bot and cls.get_nearest_lang(request.env.context.get('lang', 'en_US')))
```

Comment:
Fallback on `en_US` is usually a bad idea (even if done often) as it is possible to create db without activating en_US (see #35633)

Reply:
Done, removed it.

As mentioned in the other comment, it should be set anyway. No need of a fallback.

> request.context.lang [will read](https://github.com/odoo/odoo/blame/2e640a37d45b53d28e5f5a67eede7855bb68a1d4/odoo/http.py#L253) request.session.context.lang which is set in [setup_lang](https://github.com/odoo/odoo/blame/2e640a37d45b53d28e5f5a67eede7855bb68a1d4/odoo/http.py#L1341)